### PR TITLE
Update CMSIS-NN CONV and LSTM implementations

### PR DESCRIPTION
**Pull request by AdrianLundell**
CONV
- Set filter_dims.c to allow for use of grouped convolution
- Additionally moves all consistency checks to the prepare stage

LSTM
- Updates CMSIS-NN download SHA
- New API for arm_lstm_unidirectional_s8
- New API for arm_vector_sum_s8

BUG=#2074, bit exactness of lstm kernel.

### CODE REVIEW

File: tensorflow/lite/micro/kernels/cmsis_nn/conv.cc [hunk 1]
```diff
@@ -75,29 +75,45 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
...
+  // Bias size equal to output channels
+  if (bias != nullptr) {
+    TF_LITE_ENSURE_EQ(context, bias->dims->size, 4);
+    const int bias_size = bias->dims->data[0] * bias->dims->data[1] *
+                          bias->dims->data[2] * bias->dims->data[3];
+    TFLITE_DCHECK_EQ(bias_size, output->dims->data[3]);
+  }
 
-  // Initialize cmsis_nn input dimensions
+  // Initialize cmsis_nn dimension dimensions
```
Comment (rascani) [inline tensorflow/lite/micro/kernels/cmsis_nn/conv.cc:99]: Typo: don't need dimension twice.

File: tensorflow/lite/micro/kernels/cmsis_nn/conv.cc [hunk 2]
```diff
@@ -75,29 +75,45 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
...
+  // Input channels should be an even multiple of filter channels
+  TF_LITE_ENSURE(context, filter->dims->data[3] > 0);
+  TF_LITE_ENSURE_EQ(context, input->dims->data[3] % filter->dims->data[3], 0);
+  // Output channels should be an even multiple of the number of groups
+  const int groups = input->dims->data[3] / filter->dims->data[3];
+  TFLITE_DCHECK_EQ(output->dims->data[3] % groups, 0);
+  // Bias size equal to output channels
+  if (bias != nullptr) {
+    TF_LITE_ENSURE_EQ(context, bias->dims->size, 4);
+    const int bias_size = bias->dims->data[0] * bias->dims->data[1] *
```
Comment (rascani) [inline tensorflow/lite/micro/kernels/cmsis_nn/conv.cc:94]: You can shorten this with:

```
const int bias_size = NumElements(bias->dims);
```

**Review by rascani — COMMENTED** [review]
This LGTM, just a couple nits. I'll approve post responses.

**Comment by AdrianLundell** [comment]
Apologies for extending the original PR, due to https://github.com/ARM-software/CMSIS-NN/commit/601d96c63a4e1437d9a90c0f917684405e60d8a1 getting merged in between we want to update to match the latest version as soon as possible.

**Review by rascani — APPROVED** [review]

**Comment by rascani** [comment]
> Apologies for extending the original PR, due to [ARM-software/CMSIS-NN@601d96c](https://github.com/ARM-software/CMSIS-NN/commit/601d96c63a4e1437d9a90c0f917684405e60d8a1) getting merged in between we want to update to match the latest version as soon as possible.

Not a problem. Thanks Adrian!

**Comment by mansnils** [comment]
> > Apologies for extending the original PR, due to [ARM-software/CMSIS-NN@601d96c](https://github.com/ARM-software/CMSIS-NN/commit/601d96c63a4e1437d9a90c0f917684405e60d8a1) getting merged in between we want to update to match the latest version as soon as possible.
> 
> Not a problem. Thanks Adrian!

Thanks @rascani  for approving the PR! I can see some tests failing. Not sure why QEMU with CMSIS-NN is failing when Bluepill with CMSIS-NN is passing? And maybe the project generation error is something known?
Do we need to update the PR to get it merged?

**Comment by rascani** [comment]
> Thanks @rascani for approving the PR! I can see some tests failing. Not sure why QEMU with CMSIS-NN is failing when Bluepill with CMSIS-NN is passing? And maybe the project generation error is something known? Do we need to update the PR to get it merged?

Yeah, we'll need to get those three issues resolved for mergify to get it through.

* Cortex-M QEMU Unit tests -  I agree this seems odd. Are you able to reproduce locally?
* Code Style - You can run `tensorflow/lite/micro/tools/ci_build/test_code_style.sh --fix_formatting` to correct this (provided you have the right clang-format version installed). I can also do this for you, but I'd prefer to wait until all other issues are resolved first so that there aren't any additional code changes to land.
* project generation - No, this seems new. It looks like the script is trying to find a `LICENSE.txt` file in the root of the cmsis_nn download and it is no longer there. Was this file recently removed or renamed? We might just need to update this [line](https://github.com/tensorflow/tflite-micro/blob/main/tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc#L46).

**Comment by mansnils** [comment]
> > Thanks @rascani for approving the PR! I can see some tests failing. Not sure why QEMU with CMSIS-NN is failing when Bluepill with CMSIS-NN is passing? And maybe the project generation error is something known? Do we need to update the PR to get it merged?
> 
> Yeah, we'll need to get those three issues resolved for mergify to get it through.
> 
>     * Cortex-M QEMU Unit tests -  I agree this seems odd. Are you able to reproduce locally?
> 
>     * Code Style - You can run `tensorflow/lite/micro/tools/ci_build/test_code_style.sh --fix_formatting` to correct this (provided you have the right clang-format version installed). I can also do this for you, but I'd prefer to wait until all other issues are resolved first so that there aren't any additional code changes to land.
> 
>     * project generation - No, this seems new. It looks like the script is trying to find a `LICENSE.txt` file in the root of the cmsis_nn download and it is no longer there. Was this file recently removed or renamed? We might just need to update this [line](https://github.com/tensorflow/tflite-micro/blob/main/tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc#L46).

Thanks @rascani for the help. We will fix the three issues. At least the last two should be straight forward.

**Review by rascani — APPROVED** [review]

**Comment by mansnils** [comment]
@Mergifyio refresh

**Comment by rascani** [comment]
@Mergifyio refresh

**Pull request merged**